packages/fxa-settings/src/lib/relier-factory.ts

```diff
@@ -0,0 +1,176 @@
+const FX_DESKTOP_V3_CONTEXT = 'fx_desktop_v3'
+
+/**
+ * A react context compliant interface
+ */
+export interface IRelierContext {
+  relier: RelierData,
+  setRelier: (relier:RelierData) => void
+};
+
+/**
+ * Creation flags interface, controls the type of relier that is ultimately produced.
+ */
+export interface RelierFlags
+{
+  isDevicePairingAsAuthority():boolean;
+  isDevicePairingAsSupplicant():boolean;
+  isOauth():boolean;
+  isSyncService():boolean;
+  isV3DesktopContext():boolean;
+}
+
+/**
+ * Produces Reliers
+ */
+export class RelierFactory
+{
+  /**
+   * Produces a relier given the current context and flags extrapolated from that context.
+   * @param context A key value data store that holds the relier state
+   * @param flags A set of flags that help determine what kind of relier to produce. These flags are also based off the context.
+   * @returns A relier implementation.
+   */
+  static getRelier(
+    context:Context,
+    flags:RelierFlags
+  ) {
+
+    // Keep trying until something sticks
+    if (flags.isDevicePairingAsAuthority()) {
+      return new PairingAuthorityRelier(context);
+    }
+    if (flags.isDevicePairingAsSupplicant()) {
+      return new RelierPairingSupplicant(context);
+    }
+    if (flags.isOauth()) {
+      return new OAuthRelier(context);
+    }
+    if (flags.isSyncService() && flags.isV3DesktopContext()) {
+      return new BrowserRelier(context);
+    }
+
+    return new Relier(context);
+  }
+}
+
+
+
+/**
+ * Extrapolates flags from the state of the current context. The collective state of these flags are used by
+ * the factory to determine what underlying type of relier to create.
+ *
+ * Note: this logic was ported from fxa-content-server app-start.js.
+ */
+export class DefaultRelierFlags
+{
+  private get pathname() {
+    return this.urlContext.pathName;
+  }
+
+  constructor(
+    private urlContext:UrlContext,
+    private storageContext:StorageContext
+  ) { }
+
+  isDevicePairingAsAuthority() {
+    // this._searchParam('redirect_uri') ===
+    // Constants.DEVICE_PAIRING_AUTHORITY_REDIRECT_URI
+    return this._searchParam('redirect_uri') === DEVICE_PAIRING_AUTHORITY_REDIRECT_URI;
+  }
+
+  isDevicePairingAsSupplicant() {
+    return DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP.test(
+      this.pathname
+    );
+  }
+
+  isOauth() {
+    return (
+      !!(
+        this._searchParam('client_id') ||
+        // verification
+        this._isOAuthVerificationSameBrowser()
+      ) ||
+      this._isOAuthVerificationDifferentBrowser() ||
+      // any URL with 'oauth' in the path.
+      /oauth/.test(this.pathname)
+    );
+  }
+
+  isSyncService() {
+    return this._isService(SYNC_SERVICE)
+  }
+
+  isV3DesktopContext() {
+    return this._searchParam('context') === FX_DESKTOP_V3_CONTEXT
+  }
+
+  private _isOAuthVerificationSameBrowser() {
+    return this._isVerification() && this._isService(this._getSavedClientId());
+  }
+
+  private _isVerification() {
```

Comment:
Is it still conventional to use underscores when the method is declared `private`? I guess I see a mix of both in the codebase.

Reply:
We've decided to mostly follow the google typescript guide, which says no. I literally copied and pasted 90% of this from app-start in content server. I'll clean it up once the tests settle.